sdks/python/apache_beam/dataframe/expressions.py

```diff
@@ -48,10 +48,15 @@ def lookup(self, expr):  #  type: (Expression) -> Any
 class PartitioningSession(Session):
   """An extension of Session that enforces actual partitioning of inputs.
 
-  When evaluating an expression, inputs are partitioned according to its
-  `requires_partition_by` specifications, the expression is evaluated on each
-  partition separately, and the final result concatinated, as if this were
-  actually executed in a parallel manner.
+  Each expression is evaluated multiple times for various supported
```

Comment:
I wonder how expensive this will get. Hopefully not too bad. It could, however, mess up anything that depends on the random seed.

Reply:
Yeah I think the performance impact is worth it. We could probably reduce it by being a little more discerning in what gets re-executed. For example I don't think there's any value in doing this for expressions that have preserves=Nothing().

Really good point about the random seed. I could bracket the runs with calls to random.getstate() and random.setstate()

Reply:
Let's do that.

Reply:
Done